p2p/protocol/identify/obsaddr.go

```diff
@@ -46,13 +66,36 @@ type ObservedAddr struct {
 	LastSeen time.Time
 }
 
-func (oa *ObservedAddr) activated(ttl time.Duration) bool {
+func (oa *ObservedAddr) activated() bool {
+
 	// We only activate if other peers observed the same address
 	// of ours at least 4 times. SeenBy peers are removed by GC if
 	// they say the address more than ttl*ActivationThresh
 	return len(oa.SeenBy) >= ActivationThresh
 }
 
+func (oa *ObservedAddr) nInbound() int {
+	count := 0
+	for obs := range oa.SeenBy {
+		if oa.SeenBy[obs].connDirection == network.DirInbound {
+			count++
+		}
+	}
+
+	return count
+}
+
+func (oa *ObservedAddr) GroupKey() string {
+	for i := range addrPatterns {
+		pat := addrPatterns[i]
+		if pat.Matches(oa.Addr) {
+			return pat.String()
+		}
+	}
+
+	return ""
+}
```

Comment:
We also need to match things like `/ws`, `/quic,` etc. We should probably just call `oa.Addr.Protocols()` and concatenate the protocol codes to get the group key.

Reply:
Good point. This is done.